library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/Ac3Reader.java

```diff
@@ -188,18 +192,36 @@ private boolean skipToNextSync(ParsableByteArray pesBuffer) {
   private void parseHeader() {
     headerScratchBits.setPosition(0);
     Ac3Util.Ac3SyncFrameInfo frameInfo = Ac3Util.parseAc3SyncframeInfo(headerScratchBits);
-    if (format == null || frameInfo.channelCount != format.channelCount
-        || frameInfo.sampleRate != format.sampleRate
-        || frameInfo.mimeType != format.sampleMimeType) {
+    if ((format == null || frameInfo.sampleRate != format.sampleRate
+        || frameInfo.mimeType != format.sampleMimeType)
```

Comment:
Hm. The idea of this block was (I think) to output a new format whenever something needs to change in it. That doesn't look like it's the case any more. In particular, a change to only channelCount doesn't look like it will cause a new format to be output. I'm not sure if I'm just misunderstanding the goal, but it feels like this code should look more like:

```
Ac3Util.Ac3SyncFrameInfo frameInfo = Ac3Util.parseAc3SyncframeInfo(headerScratchBits);
int formatChannelCount = frameInfo.channelCount;
if (frameInfo.streamType == Ac3Util.STREAM_TYPE_TYPE1
    && format.channelCount == 6 && frameInfo.channelCount == 4) {
  formatChannelCount = 8;
}
if (format == null || formatChannelCount != format.channelCount
   || frameInfo.sampleRate != format.sampleRate
   || frameInfo.mimeType != format.sampleMimeType) {
  .... Create and output the format
}
```

Does that make any sense?

Reply:
Issue #2552 asked for support for the channel configuration changing during playback so it would be good to make sure this still works.

I wonder whether grouping together "I,D" into a single output sample could actually be fairly easy to do:
- It looks like we do actually now parse the header for every syncframe, which means that we know the stream type of the syncframe being read.
- At the moment we just call output.sampleMetadata when we've read sampleSize bytes, but instead we could delay calling it until we read a syncframe with stream type == 0. It would be necessary to keep track of the number of bytes read from the current syncframe and the number of bytes read for the access unit separately.
- We'd also need a way to make sure we output the last sample in the stream (as we won't read a next syncframe for this one).
- This also means we can delay the call to output.format until we've read all constituent syncframes of the access unit, which avoids the problem of outputting two formats. (The first sample would likely just get dropped in the AudioTrack when it's reconfigured with the second format, which is bad.)

I haven't tried this, so maybe something will come up that makes it impossible, but worth attempting I think. Thanks!

Reply:
andrewlewis, thanks for this idea. If you can group together "I,D" into a single output sample, that is fine.

Reply:
Whether it's ok to group "I,D" into a single output sample I think depends on whether the Dolby decoder can accept input buffers with "I,D" grouped in a single sample. Can it?

Reply:
Yes, it can. The Dolby decoder's behavior is something as below. (pseudo code)

dolby_decoder_wrapper_process(const char* p_inBuffer, const int inBufferLength, char* p_outBuffer, int* p_bytesConsumed)
{
    ......
    int timesliceComplete;
    dolby_decoder_api_addbytes(p_inBuffer, inBufferLength, p_bytesConsumed, &timesliceComplete);
    if (timeslicecomplete) { 
        dolby_decoder_api_process_timeslice(p_outBuffer);
    }
    ......
}
The first Dolby decoder API fetches the input data. As long as it gets enough data (one full timeslice and one another full I0), it will set timesliceComplete = true.
For example, in following cases, it will set timesliceComplete = true.
(1) I0, I0  (this is general 5.1 case)
(2) I0, D0, I0 (this is general 7.1 case) 
(3) I0, D0, I1, D0, I0
Once timesliceComplete is true, it will call process() API to do decoding. (Decoder will decoder the first timeslice. The second I0 just keeps in internal buffer and will be decoded next time.)  Otherwise, it just waits for more data to be inputted.

Reply:
@ybai001 It sounds like grouping "I,D" together as described above will work in that case. Do you have any plans to update the pull request to implement this? Thanks!

Reply:
@andrewlewis We have no plan to do this update recently due to the internal priority. You may close this pull request at first. (This pull request includes three parts of update. The first & third parts of update have been merged by https://github.com/google/ExoPlayer/pull/3161.) After we implement this new logic, we wil submit a new pull-request. Is it OK for you?

Reply:
Sounds good. Thanks.